src/libos/src/fs/inode_file.rs

```diff
@@ -264,6 +273,34 @@ impl INodeFile {
     pub fn abs_path(&self) -> &str {
         &self.abs_path
     }
+
+    fn check_flock_with_access_mode(&self, lock: &Flock) -> Result<()> {
+        match lock.type_ {
+            FlockType::F_RDLCK => {
+                if !self.access_mode.readable() {
+                    return_errno!(EBADF, "File not readable");
+                }
+            }
+            FlockType::F_WRLCK => {
+                if !self.access_mode.writable() {
+                    return_errno!(EBADF, "File not writable");
+                }
+            }
+            _ => (),
+        }
+        Ok(())
+    }
+}
+
+impl Drop for INodeFile {
+    fn drop(&mut self) {
+        if let Some(inode_lock_list) = self.inode.test_lock_list() {
+            if let Some(flock_list) = inode_lock_list.downcast_ref::<FlockList>() {
+                let flock = Flock::new(FlockType::F_UNLCK, 0, off_t::max_value(), true).unwrap();
+                flock_list.unlock(&flock).unwrap();
+            }
+        }
+    }
```

Comment:
After reading this `drop` method, I think I found two issues (I hope I am mistaken).

1. **A (kind-of) deadlock issue** due to the compatibility with the `spawn` syscall. If a `Arc<INodeFile>` is passed to a child process through `spawn`, then the `INodeFile::drop` method will not get invoked by the parent process (the counter of `Arc<INodeFile>` is not zero)! This means all locks on the `Arc<INodeFile>` will be inherited by the child process. But the lock's internal data structures record the pid of the parent process. And the child process cannot unlock them! This looks (kind of) like a deadlock to me.

2. **A lost-lock issue**. Consider a file opened twice in a process, resulting in two `INodeFile` instances (say `A` and `B`) backed by the same inode. Now the user first acquire some locks on `A`, then drop `B` by invoking the `close` syscall on `B`'s fd.  According the logic here, all locks acquired through `A` would suddenly get lost simply because `B` is closed. Clearly, this is not the user would expect.

Reply:
After the review, I think it could happen. The reason is that One process can unlock the flocks just at releasing the last reference to the file. So we have to find this situation when releasing the file.